text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,517 @@
+## Method shadowing
+[method-shadowing]: #method-shadowing
+
+For a smart pointer `P<T>` that implements `Deref<Target = T>`, a method call `p.m()` might call a method `P::m` on the smart pointer type itself, or it might call `T::m`. If both methods are declared, this results in an error.
+
+Rust standard library smart pointers are designed with this shadowing behavior in mind:
+
+* `Box`, `Pin`, `Rc` and `Arc` heavily use associated functions rather than methods.
+* Where they use methods, it's often with the _intention_ of shadowing a method in the inner type (e.g. `Arc::clone`).
+
+Furthermore, the `Deref` trait itself [documents this possible compatibility hazard](https://doc.rust-lang.org/nightly/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut), and the Rust API Guidelines has [a guideline about avoiding inherent methods on smart pointers](https://rust-lang.github.io/api-guidelines/predictability.html#smart-pointers-do-not-add-inherent-methods-c-smart-ptr).
+
+These method shadowing risks also apply to `Receiver`. This RFC does not make things worse for types that implement `Deref`, it only adds additional flexibility to the `self` parameter type for `T::m`.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As this feature has been cooking since 2017, many alternative implementations have been discussed.
+
+## Deref-based
+[deref-based]: #deref-based
+
+As noted in the rationale section, the currently nightly implementation implements arbitrary self types using the `Deref` trait.
+
+## No blanket implementation for `Deref`
+[no-blanket-implementation]: #no-blanket-implementation
+
+The other major approach previously discussed is to have a `Receiver` trait, as proposed in this RFC, but without a blanket implementation for `T: Deref`. Blanket implementations are unusual for core Rust traits, but the authors of this RFC believe it's necessary in this case.
+
+Specifically, this RFC proposes that the existing method search algorithm is modified to search the `Receiver` chain _instead of_ the `Deref` chain.
+
+It's therefore a major compatibility break if existing `Deref` implementors cease to be usable as `self` parameters. Just in the standard library, we'd have to add `Receiver` implementations for `Cow`, `Ref`, `ManuallyDrop` and possibly many other existing implementors of `Deref`: third party libraries would have to do the same. Without that, method calls on these types would not be possible:
+
+```rust
+fn main() {
+    let ref_cell = RefCell::new(/* something cloneable */);
+    ref_cell.borrow().clone(); // no longer possible if:
+        // 1) we cease to explore Deref in identifying method candidates
+        // 2) Ref doesn't implement Receiver.
+}
+```
+
+This doesn't just break people previously using the unstable Rust `arbitrary_self_type` feature; it breaks stable Rust usages as well. Obviously this is not acceptable, so we believe the blanket implementation is necessary.
```

Comment:
> In this case though the authors of this RFC believe it necessary, because the RFC proposes that the existing method search algorithm is modified to search the `Receiver` chain _instead of_ the `Deref` chain, so it would be a major compatibility break if existing `Deref` implementors would suddenly be required to also implement `Receiver` as well to get the current behaviour.

But... it still *doesn't* search the receiver chain, right? Like you said, resolution first looks through the deref chain, and then for each method, looks if the receiver type is allowed, based on the receiver trait?

Reply:
Yeah, I wrote this before I wrote https://github.com/rust-lang/rfcs/pull/3519#discussion_r1400808795, this will need some modifications after that.

Reply:
@madsmtm In theory, it *would* be possible to introduce a mechanism that'd allow introducing new supertraits for a trait, without breaking backwards compatibility.

Imagine if we had something like "inherent supertraits", such that we could say the `Target` type in `Deref` is actually part of `Receiver`, and *if* you write just an `impl Deref for T` that includes `Target` then that implicitly defines `Receiver`, while you also have the option of writing an explicit `impl Receiver for T` in which case you can't also define `Target` when defining `impl Deref for T`.

Effectively, this would give the option of "splitting out" some methods or associated types of a trait into a supertrait, compatibly, such that anyone who acts like they're *not* split out is still correct, and anyone who acts like they *are* split out is also correct.

This wouldn't be worth it for `Receiver`/`Deref` alone, but it seems like a potentially useful general feature. The question is then whether we think that would be *better* than the blanket impl in any way. Is there anything we could do given the supertrait approach (`Deref: Receiver`) that we *couldn't* do with the blanket impl?

Reply:
> This wouldn't be worth it for `Receiver`/`Deref` alone, but it seems like a potentially useful general feature. The question is then whether we think that would be better than the blanket impl in any way. Is there anything we could do given the supertrait approach (`Deref: Receiver`) that we couldn't do with the blanket impl?

I don't think the supertrait approach gives us anything other than being more in line with how things currently work.

(E.g. to implement `PartialOrd`, you have to implement `PartialEq` first, instead of e.g. `PartialEq` having a blanket impl for `T: PartialOrd`).